migration/1703398409668-add_missed_op_donations_to_db.ts

```diff
@@ -0,0 +1,246 @@
+    transactionNetworkId: NETWORK_IDS.OPTIMISTIC,
+    createdAt: millisecondTimestampToDate(1703348735000),
+  },
+
+  // https://optimistic.etherscan.io/tx/0x2e410bf8b53bbefe6025a6409fc9cee4f29e79b7f3025f260d85fd7555496e46
+  {
+    fromWalletAddress: '0x41cb654d1f47913acab158a8199191d160dabe4a',
+    toWalletAddress: '0x29185eB8cfD22Aa719529217bFbadE61677e0Ad2',
+    // https://giveth.io/project/digital-street-musician-doing-free-live-concerts
+    projectId: 2961,
+    transactionId:
+      '0x2e410bf8b53bbefe6025a6409fc9cee4f29e79b7f3025f260d85fd7555496e46',
+    currency: 'USDC',
+    tokenAddress: '0x7F5c764cBc14f9669B88837ca1490cCa17c31607',
+    amount: 4,
+    valueUsd: 4,
+    transactionNetworkId: NETWORK_IDS.OPTIMISTIC,
+    createdAt: millisecondTimestampToDate(1703349653000),
+  },
+];
+
+export class addMissedOpDonationsToDb1703398409668
+  implements MigrationInterface
+{
+  async up(queryRunner: QueryRunner): Promise<void> {
+    const environment = config.get('ENVIRONMENT') as string;
+
+    if (environment !== 'production') {
+      // tslint:disable-next-line:no-console
+      console.log('We just want to create these donations in production DB');
+      return;
+    }
+
+    await AppDataSource.initialize();
+    for (const tx of transactions) {
+      const user = (
+        await queryRunner.query(`SELECT * FROM public.user
+        WHERE lower("walletAddress")=lower('${tx.fromWalletAddress}')`)
+      )[0];
+
+      // Set true for isTokenEligibleForGivback, isProjectVerified because Ashley mentioned we want to pay givback for them
+      const createdAt = moment(tx.createdAt).format('YYYY-MM-DD HH:mm:ss');
+      const project = (await findProjectById(
+        tx.projectId as number,
+      )) as Project;
+
+      const { givbackFactor, projectRank, powerRound, bottomRankInRound } =
+        await calculateGivbackFactor(tx.projectId as number);
+
+      await queryRunner.query(`
+           INSERT INTO donation ("toWalletAddress", "projectId", "fromWalletAddress", "userId", amount, currency, "transactionId", "transactionNetworkId", anonymous, "valueUsd", status,
+            "segmentNotified", "isTokenEligibleForGivback", "isProjectVerified", "createdAt", "givbackFactor", "powerRound", "projectRank", "bottomRankInRound", "qfRoundId", "tokenAddress")
+           VALUES ('${tx.toWalletAddress}', ${tx.projectId}, '${tx.fromWalletAddress}', ${user.id}, ${tx.amount}, '${tx.currency}', '${tx.transactionId}', ${tx.transactionNetworkId}, false, ${tx.valueUsd}, 'verified',
+             true, true, true, '${createdAt}', ${givbackFactor}, ${powerRound}, ${projectRank}, ${bottomRankInRound}, ${QF_ROUND_ID}, '${tx.tokenAddress}');
+                `);
+
+      await updateUserTotalDonated(user.id);
+      await updateUserTotalReceived(project.adminUser?.id);
+      await updateTotalDonationsOfProject(tx.projectId as number);
+      await refreshProjectEstimatedMatchingView();
+      await refreshProjectDonationSummaryView();
```

Comment:
These two can be done out of loop and only once

Reply:
Thanks @aminlatifi , you are right, I moved it to out of the function